go-controller/pkg/ovn/egressgw.go

```diff
@@ -39,6 +41,114 @@ type ovnRoute struct {
 	shouldExist bool
 }
 
+type externalRouteInfo struct {
+	sync.RWMutex
+	// podExternalRoutes is a cache keeping the LR routes added to the GRs when
+	// the k8s.ovn.org/routing-external-gws annotation is used. The first map key
+	// is the podIP, the second the GW and the third the GR
+	routes map[string]map[string]string
+}
+
+func (oc *Controller) ensureRouteInfoLocked(podName ktypes.NamespacedName, readOnly bool) (*externalRouteInfo, func(), error) {
+	oc.egressGWCacheMutex.Lock()
+	routeInfo := oc.egressGWCache[podName]
+	routeInfoExisted := false
+	if routeInfo == nil {
+		routeInfo = &externalRouteInfo{
+			routes: make(map[string]map[string]string),
+		}
+		// we are creating routeInfo and going to set it in podExternalRoutes map
+		// so safe to hold the lock while we create and add it
+		defer oc.egressGWCacheMutex.Unlock()
+		oc.egressGWCache[podName] = routeInfo
+	} else {
+		routeInfoExisted = true
+		// if we found and existing nsInfo, do not hold the namespaces lock
+		// while waiting for nsInfo to Lock
+		oc.egressGWCacheMutex.Unlock()
+	}
+
+	var unlockFunc func()
+	if readOnly {
+		unlockFunc = routeInfo.RUnlock
+		routeInfo.RLock()
+	} else {
+		unlockFunc = routeInfo.Unlock
+		routeInfo.Lock()
+	}
+
+	if routeInfoExisted {
+		// Check that the namespace wasn't deleted while we were waiting for the lock
+		oc.egressGWCacheMutex.Lock()
+		defer oc.egressGWCacheMutex.Unlock()
+		if routeInfo != oc.egressGWCache[podName] {
```

Comment:
as you can tell from the comment I just copied this logic from nsInfo :)

How this works is like this:
1. We dont want to hold the overall cache lock, while we try to lock the routeInfo. This can lead to dead lock.
2. We grab the cache lock for a short time, just to find which routeInfo is currently in the cache.
3. We unlock the cache, and try to lock routeInfo. This frees up the cache, so that access blocked to a single routeInfo wouldn't block everyone else from accessing the cache.
4. Once we get a lock  on routeInfo, we dont know that whatever was accessing it may have altered the cache for that key.
5. So now we lock the cache (although looking at the code now this could be an Rlock, probably doesnt really matter).
6. Now we compare the *pointer* of routeInfo to the value in the cache. That way we know if the cache entry for this now points to something else, or nil. In that case we know the value we are referencing is stale and we should abandon ensuring it.

Reply:
FTR, I think this algorithm is pretty non-obvious and we gotta have some way to fix it. Refcounting, or something else that's clearer, cause this keeps coming up...

Reply:
i added a bunch of comments inside of the function